jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -1648,4 +1590,155 @@ private void complete()
+            }
+        }
+
+        private int reserveSlot(Slot slot, int streamId)
+        {
+            if (streamId <= 0)
+            {
+                synchronized (this)
+                {
+                    streamId = localStreamIds.getAndAdd(2);
+                    slots.offer(slot);
+                }
+            }
+            else
+            {
+                synchronized (this)
+                {
+                    slots.offer(slot);
+                }
+            }
+            return streamId;
+        }
+
+        private void createLocalStream(Slot slot, Frame frame, Promise<Stream> promise, Stream.Listener listener, int streamId)
+        {
+            IStream stream = HTTP2Session.this.createLocalStream(streamId);
+            stream.setListener(listener);
+            slot.entry = new ControlEntry(frame, stream, new StreamPromiseCallback(promise, stream));
+            flush();
+        }
+
+        private void freeSlotAndFailPromise(Slot slot, Promise<Stream> promise, Throwable x)
+        {
+            synchronized (this)
+            {
+                slots.remove(slot);
+            }
+            flush();
+            promise.failed(x);
+        }
+
+        /**
+         * Flush goes over the entries of the slots queue to flush the entries,
+         * until either one of the following two conditions is true:
+         *  - The queue is empty.
+         *  - It reaches a slot with a null entry.
+         *  When a slot with a null entry is encountered, this means a concurrent thread reserved a slot
+         *  but hasn't set its entry yet. Since entries must be flushed in order, the thread encountering
+         *  the null entry must bail out and it is up to the concurrent thread to finish up flushing.
+         *  Note that only one thread can flush at any one time, if two threads happen to call flush
+         *  concurrently, one will do the work while the other will bail out, so it is safe that all
+         *  threads call flush after they're done reserving a slot and setting the entry.
+         */
+        private void flush()
+        {
+            Thread thread = Thread.currentThread();
+            boolean queued = false;
+            while (true)
+            {
+                ControlEntry entry = null;
+                synchronized (this)
+                {
+                    if (flushing != null && flushing != thread)
+                        return;
+                    flushing = thread;
+
+                    Slot slot = slots.peek();
+                    if (slot != null)
+                        entry = slot.entry;
+                    if (entry == null)
+                    {
+                        flushing = null;
+                        break;
+                    }
+                    slots.poll();
+                }
+                queued |= flusher.append(entry);
+            }
+            if (queued)
+                flusher.iterate();
+        }
```

Comment:
There is a mixed style in this look that makes it hard to read.  I also think that perhaps there is a bug in the different handling between peeking a null slot (no more slots) and seeing a slot with a null entry (another thread will come and finish flushing).   Perhaps the loop should be:
```java
            while (true)
            {
                ControlEntry entry = null;
                synchronized (this)
                {
                    if (flushing == null)
                        flushing = thread;
                    else if (flushing != thread)
                        return; // another thread is flushing

                    Slot slot = slots.peek();
                    if (slot == null)
                        break; // No more slots, so we may iterate on the flusher
                        
                    entry = slot.entry;
                    if (entry == null)
                    {
                        flushing = null;
                        return; // Another thread will finish the flush
                    }
                    slots.poll();
                }
                queued |= flusher.append(entry);
            }
            if (queued)
                flusher.iterate();
        }
```

Reply:
I agree on the intent of adding a comment about why we exit the loop.

However, the case where `slot==null` must set `flushing=null` before `break`ing.

And the case `entry==null` must `break`, not `return`, to iterate over previous slots that may have been queued to flush.

Reply:
I like the extra comments explaining the reasons for exiting the loop and the fact that the `entry` variable doesn't have to be initialized to null (despite the fact that your snippet still initializes it). Let me update the PR with those changes.